File: python/DUE/System.py
from enum import Enum
import time

class SystemController:
    class ResetOption(Enum):
        SystemReset = 0
        Bootloader = 1

    def __init__(self, serialPort):
        self.serialPort = serialPort

    def Reset(self, option : Enum):
        cmd = "reset({0})".format(1 if option.value == 1 else 0)
        self.serialPort.WriteCommand(cmd)
        # The device will reset in bootloader or system reset
        self.serialPort.Disconnect()

    def GetTickMicroseconds(self):
        cmd = "print(tickus())"
        self.serialPort.WriteCommand(cmd)
        res = self.serialPort.ReadRespone()
        if res.success:
            try:
                return int(res.respone)
            except:
                pass
        return -1
    
    def GetTickMilliseconds(self):
        cmd = "print(tickms())"
        self.serialPort.WriteCommand(cmd)
        res = self.serialPort.ReadRespone()
        if res.success:
            try:
                return int(res.respone)
            except:
                pass
        return -1
    
    def IsProgramMode(self)->bool:
        cmd = bytearray(1)
        data = bytearray(1)
    
        cmd[0] = 255
        
        self.serialPort.WriteRawData(cmd, 0, 1)

        time.sleep(0.01)

        self.serialPort.ReadRawData(data,0,1)

        if (data[0]== 255):
            return True
        
        return False
    
    def ExitProgramMode(self) -> bool:
        cmd = bytearray(1)
        cmd[0] = 127

        self.serialPort.WriteRawData(cmd, 0, 1)

        time.sleep(0.01)

        res = self.serialPort.ReadRespone()

        return res.success

File: dotnet/GHIElectronics.Due_/System.cs
using System;
using System.Collections.Generic;
using System.Diagnostics.Contracts;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace GHIElectronics.DUE {

    public partial class DUEController {

        public enum ResetOption {
            SystemReset = 0,
            Bootloader
            

        }
        public class SystemController {

            SerialInterface serialPort;

            public SystemController(SerialInterface serialPort) => this.serialPort = serialPort;

            public void Reset(ResetOption option) {

                var cmd = string.Format("reset({0})", option == ResetOption.Bootloader ? 1 : 0 );
                this.serialPort.WriteCommand(cmd);

                // The device will reset in bootloader or system reset
                this.serialPort.Disconnect();

            }

            public int GetTickMicroseconds() {
                var cmd = string.Format("print(tickus())");

                this.serialPort.WriteCommand(cmd);

                var res = this.serialPort.ReadRespone();

                if (res.success) {
                    try {
                        var tick = int.Parse(res.respone);
                        return tick;
                    }
                    catch {

                    }

                }

                return -1;
            }

            public int GetTickMilliseconds() {
                var cmd = string.Format("print(tickms())");

                this.serialPort.WriteCommand(cmd);

                var res = this.serialPort.ReadRespone();

                if (res.success) {
                    try {
                        var tick = int.Parse(res.respone);
                        return tick;
                    }
                    catch {

                    }

                }

                return -1;
            }

            
        }
    }
}
